File: src/runtime/storage/tests/ramdisk-direct-store-muxer-integration-test.ts
/**
 * @license
 * Copyright (c) 2019 Google Inc. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {assert} from '../../../platform/chai-web.js';
import {ProxyMessageType, ProxyMessage} from '../store-interface.js';
import {RamDiskStorageKey} from '../drivers/ramdisk.js';
import {Exists} from '../drivers/driver.js';
import {DirectStoreMuxer} from '../direct-store-muxer.js';
import {EntityType, MuxType} from '../../../types/lib-types.js';
import {Manifest} from '../../manifest.js';
import {CRDTMuxEntity} from '../storage.js';
import {Identified, CRDTEntity, EntityOpTypes, CRDTSingleton} from '../../../crdt/lib-crdt.js';
import {StoreInfo} from '../store-info.js';

function assertHasModel(message: ProxyMessage<CRDTMuxEntity>, model: CRDTEntity<Identified, Identified>) {
  if (message.type === ProxyMessageType.ModelUpdate) {
    assert.deepEqual(message.model, model.getData());
  } else {
    assert.fail('message is not a ModelUpdate');
  }
}

describe('RamDisk + Direct Store Muxer Integration', async () => {
  it('will allow storage of a number of objects', async () => {
    const manifest = await Manifest.parse(`
      schema Simple
        txt: Text
    `);
    const simpleSchema = manifest.schemas.Simple;

    const storageKey = new RamDiskStorageKey('unique');
    const type = new MuxType(new EntityType(simpleSchema));
    const storeInfo = new StoreInfo<MuxType<EntityType>>({
        storageKey, exists: Exists.ShouldCreate, type, id: 'base-store-id'});
    const store = await DirectStoreMuxer.construct<Identified, Identified, CRDTMuxEntity>({
      storageKey,
      exists: Exists.ShouldCreate,
      type: new MuxType(new EntityType(simpleSchema)),
      storeInfo,
    });


    const entity1 = new CRDTEntity({txt: new CRDTSingleton<{id: string, value: string}>()}, {});
    entity1.applyOperation({type: EntityOpTypes.Set, field: 'txt', value: {id: 'text1', value: 'text1'}, actor: 'me', versionMap: {'me': 1}});

    const entity2 = new CRDTEntity({txt: new CRDTSingleton<{id: string, value: string}>()}, {});
    entity2.applyOperation({type: EntityOpTypes.Set, field: 'txt', value: {id: 'text2', value: 'text2'}, actor: 'me', versionMap: {'me': 1}});

    const id = store.on(async (message) => {return;});
    await store.onProxyMessage({type: ProxyMessageType.ModelUpdate, model: entity1.getData(), id, muxId: 'thing0'});
    await store.onProxyMessage({type: ProxyMessageType.ModelUpdate, model: entity2.getData(), id, muxId: 'thing1'});

    await store.idle();

    let message: ProxyMessage<CRDTMuxEntity>;
    let muxId: string;
    const id2 = store.on(async (m) => {message = m; muxId = m.muxId;});
    await store.onProxyMessage({type: ProxyMessageType.SyncRequest, id: id2, muxId: 'thing0'});
    assertHasModel(message, entity1);
    assert.strictEqual(muxId, 'thing0');
    await store.onProxyMessage({type: ProxyMessageType.SyncRequest, id: id2, muxId: 'thing1'});
    assertHasModel(message, entity2);
    assert.strictEqual(muxId, 'thing1');
    await store.onProxyMessage({type: ProxyMessageType.SyncRequest, id: id2, muxId: 'not-a-thing'});
    assertHasModel(message, new CRDTEntity({txt: new CRDTSingleton<{id: string, value: string}>()}, {}));
    assert.strictEqual(muxId, 'not-a-thing');
  });
});

File: javatests/arcs/core/storage/RamDiskDirectStoreMuxerIntegrationTest.kt
/*
 * Copyright 2019 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.storage

import arcs.core.crdt.CrdtCount
import arcs.core.crdt.CrdtCount.Operation.Increment
import arcs.core.crdt.CrdtCount.Operation.MultiIncrement
import arcs.core.crdt.CrdtData
import arcs.core.crdt.CrdtModel
import arcs.core.crdt.CrdtOperation
import arcs.core.crdt.CrdtOperationAtTime
import arcs.core.data.CountType
import arcs.core.storage.driver.RamDisk
import arcs.core.storage.driver.RamDiskDriverProvider
import arcs.core.storage.keys.RamDiskStorageKey
import arcs.core.storage.testutil.testWriteBackProvider
import com.google.common.truth.Truth.assertThat
import com.google.common.truth.Truth.assertWithMessage
import kotlinx.atomicfu.atomic
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.Job
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.runBlockingTest
import org.junit.After
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

/** Tests for [RamDiskDriver] coupled with [DirectStoreMuxer]. */
@ExperimentalCoroutinesApi
@RunWith(JUnit4::class)
class RamDiskDirectStoreMuxerIntegrationTest {

  @Before
  fun setup() {
    DefaultDriverFactory.update(RamDiskDriverProvider())
  }

  @After
  fun teardown() = runBlocking {
    RamDisk.clear()
  }

  @Suppress("UNCHECKED_CAST")
  @Test
  fun allowsStorageOf_aNumberOfObjects() = runBlockingTest {
    val message = atomic<ProxyMessage<CrdtCount.Data, CrdtCount.Operation, Int>?>(null)
    val muxId = atomic<String?>(null)
    var job = Job()

    val storageKey = RamDiskStorageKey("unique")
    val store = DirectStoreMuxerImpl<CrdtData, CrdtOperationAtTime, Any?>(
      storageKey = storageKey,
      backingType = CountType(),
      coroutineScope = this,
      writeBackProvider = ::testWriteBackProvider,
      devTools = null
    ).also {
      it.on { muxedProxyMessage ->
        message.value = muxedProxyMessage.message
          as ProxyMessage<CrdtCount.Data, CrdtCount.Operation, Int>
        muxId.value = muxedProxyMessage.muxId
        job.complete()
      }
    }

    val count1 = CrdtCount()
    count1.applyOperation(Increment("me", version = 0 to 1))

    val count2 = CrdtCount()
    count2.applyOperation(MultiIncrement("them", version = 0 to 10, delta = 15))

    store.onProxyMessage(
      MuxedProxyMessage<CrdtData, CrdtOperationAtTime, Any?>(
        "thing0",
        ProxyMessage.ModelUpdate(count1.data, null)
      )
    )
    store.onProxyMessage(
      MuxedProxyMessage<CrdtData, CrdtOperationAtTime, Any?>(
        "thing1",
        ProxyMessage.ModelUpdate(count2.data, null)
      )
    )

    store.idle()

    store.onProxyMessage(
      MuxedProxyMessage(
        "thing0",
        ProxyMessage.SyncRequest(null)
      )
    )
    job.join()
    message.value.assertHasData(count1)
    assertThat(muxId.value ?: "huh, it was null.").isEqualTo("thing0")

    message.value = null
    muxId.value = null
    job = Job()
    store.onProxyMessage(
      MuxedProxyMessage(
        "thing1",
        ProxyMessage.SyncRequest(null)
      )
    )
    job.join()
    message.value.assertHasData(count2)
    assertThat(muxId.value ?: "huh, it was null.").isEqualTo("thing1")

    message.value = null
    muxId.value = null
    job = Job()
    store.onProxyMessage(
      MuxedProxyMessage(
        "not-a-thing",
        ProxyMessage.SyncRequest(null)
      )
    )
    job.join()
    message.value.assertHasData(CrdtCount())
    assertThat(muxId.value ?: "huh, it was null.").isEqualTo("not-a-thing")
  }

  private fun <Data, Op, ConsumerData> ProxyMessage<Data, Op, ConsumerData>?.assertHasData(
    expectedModel: CrdtModel<Data, Op, ConsumerData>
  ) where Data : CrdtData, Op : CrdtOperation {
    assertWithMessage("Message must be initialized.").that(this).isNotNull()
    when (this) {
      is ProxyMessage.ModelUpdate -> assertThat(model).isEqualTo(expectedModel.data)
      else -> fail("Message $this is not a ModelUpdate")
    }
  }
}
